File: apps/api/src/insightgraph_api/routes/query.py
from __future__ import annotations

from fastapi import APIRouter, HTTPException, Request

from insightgraph_graph.reader import GraphReader

router = APIRouter(prefix="/api/v1", tags=["query"])


def _get_reader(request: Request) -> GraphReader:
    return GraphReader(request.app.state.neo4j)


@router.get("/entities/search")
async def search_entities(
    request: Request,
    q: str = "",
    type: str | None = None,
    limit: int = 50,
) -> dict:
    """Search for entities by name and optional type filter."""
    reader = _get_reader(request)
    entities = await reader.find_entities(name=q or None, entity_type=type, limit=limit)
    return {"entities": entities, "count": len(entities)}


@router.get("/entities/{entity_id}")
async def get_entity(request: Request, entity_id: str) -> dict:
    """Get a specific entity by ID."""
    reader = _get_reader(request)
    entity = await reader.get_entity(entity_id)
    if not entity:
        raise HTTPException(404, f"Entity {entity_id} not found")
    return entity


@router.get("/entities/{entity_id}/claims")
async def get_entity_claims(request: Request, entity_id: str) -> dict:
    """Get all claims about an entity."""
    reader = _get_reader(request)
    entity = await reader.get_entity(entity_id)
    if not entity:
        raise HTTPException(404, f"Entity {entity_id} not found")
    entity_name = entity.get("canonical_name") or entity.get("name", "")
    claims = await reader.get_claims_about(entity_name)
    return {"entity_id": entity_id, "claims": claims, "count": len(claims)}


@router.get("/entities/{entity_id}/metrics")
async def get_entity_metrics(
    request: Request,
    entity_id: str,
    metric_name: str | None = None,
) -> dict:
    """Get metric values for an entity."""
    reader = _get_reader(request)
    entity = await reader.get_entity(entity_id)
    if not entity:
        raise HTTPException(404, f"Entity {entity_id} not found")
    entity_name = entity.get("canonical_name") or entity.get("name", "")
    metrics = await reader.get_metric_history(metric_name or "", entity_name)
    return {"entity_id": entity_id, "metrics": metrics, "count": len(metrics)}


@router.get("/claims/{claim_id}/evidence")
async def get_claim_evidence(request: Request, claim_id: str) -> dict:
    """Get the source evidence for a claim."""
    reader = _get_reader(request)
    evidence = await reader.find_evidence_for_claim(claim_id)
    return {"claim_id": claim_id, "evidence": evidence, "count": len(evidence)}


@router.get("/subgraph/question")
async def get_subgraph(request: Request, q: str = "") -> dict:
    """Retrieve a relevant subgraph for a question."""
    if not q:
        raise HTTPException(400, "Query parameter 'q' is required")
    reader = _get_reader(request)
    # Find entities matching the question, then get subgraph
    entities = await reader.find_entities(name=q, limit=3)
    if not entities:
        return {"nodes": [], "edges": []}
    top_id = entities[0].get("entity_id", "")
    return await reader.get_subgraph(top_id, depth=2)


@router.get("/reports")
async def list_reports(request: Request) -> dict:
    """List all ingested reports."""
    reader = _get_reader(request)
    reports = await reader.list_reports()
    return {"reports": reports, "count": len(reports)}


@router.get("/reports/{report_id}")
async def get_report(request: Request, report_id: str) -> dict:
    """Get a specific report by ID."""
    reader = _get_reader(request)
    report = await reader.get_report(report_id)
    if not report:
        raise HTTPException(404, f"Report {report_id} not found")
    return report

File: apps/api/src/routes/query.ts
import { Hono } from "hono";
import type { AppState } from "../app";
import { parseIntParam } from "../lib/validators";

export const queryRoutes = new Hono<AppState>();

queryRoutes.get("/entities/search", async (c) => {
  const reader = c.get("store").reader();
  const name = c.req.query("name");
  const entityType = c.req.query("entity_type");
  const limit = parseIntParam("limit", c.req.query("limit"), {
    min: 1,
    max: 1000,
    default: 50,
  });

  const results = await reader.findEntities(name, entityType, limit);
  return c.json(results);
});

queryRoutes.get("/entities/:entityId", async (c) => {
  const reader = c.get("store").reader();
  const entity = await reader.getEntity(c.req.param("entityId"));
  if (!entity) return c.json({ error: "Entity not found" }, 404);
  return c.json(entity);
});

queryRoutes.get("/entities/:entityId/claims", async (c) => {
  const reader = c.get("store").reader();
  // entityId here is actually entity_name in the Python API
  const results = await reader.getClaimsAbout(c.req.param("entityId"));
  return c.json(results);
});

queryRoutes.get("/entities/:entityId/metrics", async (c) => {
  const reader = c.get("store").reader();
  const results = await reader.getEntityMetrics(c.req.param("entityId"));
  return c.json(results);
});

queryRoutes.get("/claims/:claimId/evidence", async (c) => {
  const reader = c.get("store").reader();
  const results = await reader.findEvidenceForClaim(c.req.param("claimId"));
  return c.json(results);
});

queryRoutes.get("/subgraph/question", async (c) => {
  const reader = c.get("store").reader();
  const nodeId = c.req.query("node_id");
  const depth = parseIntParam("depth", c.req.query("depth"), {
    min: 1,
    max: 5,
    default: 2,
  });
  if (!nodeId) return c.json({ error: "node_id required" }, 400);
  const result = await reader.getSubgraph(nodeId, depth);
  return c.json(result);
});

queryRoutes.get("/reports", async (c) => {
  const reader = c.get("store").reader();
  const reports = await reader.listReports();
  return c.json(reports);
});

queryRoutes.get("/reports/:reportId", async (c) => {
  const reader = c.get("store").reader();
  const report = await reader.getReport(c.req.param("reportId"));
  if (!report) return c.json({ error: "Report not found" }, 404);
  return c.json(report);
});

/**
 * Get the full subgraph for one report (Report node + all connected
 * entities/claims/metrics). Implemented with backend-specific code paths:
 *   - Neo4j: runs a tailored Cypher query (original behavior preserved).
 *   - SQLite: uses GraphReader.getSubgraph, which walks via a recursive CTE.
 *
 * On SQLite we still honor the `depth` query parameter.
 */
queryRoutes.get("/reports/:reportId/graph", async (c) => {
  const store = c.get("store");
  const reportId = c.req.param("reportId");
  const depth = parseIntParam("depth", c.req.query("depth"), {
    min: 1,
    max: 5,
    default: 3,
  });

  if (store.kind !== "neo4j") {
    // Non-Neo4j backends (SQLite / FalkorDB) don't run the bespoke Cypher
    // query below — SQLite because it doesn't speak Cypher, FalkorDB because
    // some Neo4j-only functions (`elementId`) aren't available. Both backends
    // provide the same { nodes, edges } shape via the generic getSubgraph
    // reader method, so we use that as a portable fallback.
    const reader = store.reader();
    const result = await reader.getSubgraph(reportId, depth);
    return c.json(result);
  }

  const neo4j = c.get("neo4j");
  if (!neo4j) {
    return c.json(
      { error: `Raw Cypher endpoint not available on backend '${store.kind}'` },
      501,
    );
  }

  // Custom query that walks from the Report node through structure + semantic edges.
  // We intentionally skip the huge SourceSpan/Paragraph fan-out by default so the
  // client gets a readable graph focused on entities, claims, metrics.
  const cypher = `
    MATCH (r:Report {report_id: $reportId})
    OPTIONAL MATCH (r)<-[:SOURCED_FROM]-(e:Entity)
    OPTIONAL MATCH (e)<-[:ABOUT|MENTIONS]-(c:Claim)
    OPTIONAL MATCH (e)-[:HAS_VALUE]->(mv:MetricValue)-[:MEASURES]->(m:Metric)
    OPTIONAL MATCH (e)-[rel]-(e2:Entity) WHERE e2 <> e AND (e2)-[:SOURCED_FROM]->(r)
    WITH collect(DISTINCT r) + collect(DISTINCT e) + collect(DISTINCT c) + collect(DISTINCT m) + collect(DISTINCT mv) + collect(DISTINCT e2) AS nodeList,
         collect(DISTINCT rel) AS entityRels
    UNWIND nodeList AS node
    WITH collect(DISTINCT {
      id: elementId(node),
      labels: labels(node),
      props: properties(node)
    }) AS nodes, entityRels
    MATCH (r2:Report {report_id: $reportId})
    OPTIONAL MATCH p = (r2)-[*1..${depth}]-(other)
    WHERE any(n IN nodes WHERE n.id = elementId(other))
    UNWIND relationships(p) AS rel
    WITH nodes, entityRels, collect(DISTINCT rel) AS pathRels
    WITH nodes, entityRels + pathRels AS allRels
    UNWIND allRels AS rel
    WITH nodes, collect(DISTINCT {
      id: elementId(rel),
      type: type(rel),
      startId: elementId(startNode(rel)),
      endId: elementId(endNode(rel)),
      props: properties(rel)
    }) AS edges
    RETURN nodes, edges
  `;

  const session = neo4j.session();
  try {
    const result = await session.run(cypher, { reportId });
    const record = result.records[0];
    if (!record) {
      return c.json({ nodes: [], edges: [] });
    }
    const { toPlainObject } = await import("@insightgraph/graph");
    return c.json({
      nodes: toPlainObject(record.get("nodes")),
      edges: toPlainObject(record.get("edges")),
    });
  } finally {
    await session.close();
  }
});
